Code verification tutorial for TMS (#18281)
idaholab/moose: tutorials/tutorial03_verification/doc/content/tutorial03_mms.md

# Method of Manufactured Solution

!---

## Summary

In practice an exact solution is not known, since there would be no reason to perform a
simulation if the answer is known.

It is possible to "manufacture" a solution using the [!ac](MMS).

The two-dimensional heat equation will be used to simulate a physical system and used as basis
for spatial and temporal convergence study using the [!ac](MMS).

!---

## Theory: Method of Manufactured Solution

The[!ac](MMS) can be summarized into three basic steps:

1. Assume a solution for the [!ac](PDE) of interest: $T=2tx^2$

2. Apply assumed solution to compute a residual function. For example,

   !equation
   \begin{align*}
   \frac{\partial T}{\partial x} + t &= 0 \\
   \frac{\partial}{\partial x}(2tx^2) + t &= 4tx
   \end{align*}

3. Apply the negative of the computed function and perform the desired convergence study. For example,
   the example [!ac](PDE) above becomes:

   !equation
   \frac{\partial T}{\partial x} + t - 4tx = 0

   This [!ac](PDE) has a known solution, $T=2tx^2$, that can be used to compute the error and
   perform a convergence study.

!---

### Assumed Solution

> The form of the solution is arbitrary, but certain characteristics are desirable to eliminate the
> accumulation of error unnecessarily.

Reviewer: ```suggestion
The form of the solution is arbitrary, but certain characteristics are desirable to eliminate an unnecessary
accumulation of error.
```

!---

### Domain

The problem requires a two-dimensional rectangular domain, which can be defined using
the Mesh System as defined in the `[Mesh]` block  of the input.

!listing tutorial03_verification/step04_mms/2d_main.i link=False block=Mesh

!---

### Variable

There is a single unknown, temperature ($T$), to compute. This unknown is declared using the
Variables System in the `[Variables]` block and used the default
configuration of a first-order Lagrange finite element variable.

!listing tutorial03_verification/step04_mms/2d_main.i link=False block=Variables

!---

### Initial Conduction

The initial condition ($T_0$) is applied using the Initial Condition System
in the `[ICs]` block. In this case a constant value throughout the domain is provided.

!listing tutorial03_verification/step04_mms/2d_main.i link=False block=ICs

!---

### Heat Source

The volumetric heat source, $\dot{q}$, is defined using the
Function System using the `[Functions]` block. The variables "x", "y", "z", and "t" are always
available and represent the corresponding spatial location and time.

!listing tutorial03_verification/step04_mms/2d_main.i link=False block=Functions

!---

### Kernels

> The "volumetric" portions equation weak form are defined using the

Reviewer: ```suggestion
The "volumetric" portions of the weak form of the equation are defined using the
```

Assume a solution to the desired [!ac](PDE).

!equation
T = x\cdot y\cdot\textrm{exp}(-1/32400 t)

- The solution can be exactly represented by the first-order shape functions, thus
  does not result in spatial error accumulation.
- The first-order time integration cannot capture the time portion of the equation exactly.

!---

### Temporal Convergence: Forcing Function

The `mms` package can compute the necessary forcing function and output the the input file syntax
for both the forcing function and the assumed solution.

!listing tutorial03_verification/step04_mms/step04_function.py link=false start=MooseDocs:start:temporal end=MooseDocs:end:temporal include-start=0

!---

Executing this script, assuming a name of `temporal_function.py` results in the following output.

!listing
$ python temporal_function.py
[mms_force]
  type = ParsedFunction
  value = '-3.08641975308642e-5*x*y*cp*rho*exp(-3.08641975308642e-5*t) - shortwave*exp(y*kappa)*sin((1/2)*x*pi)*sin((1/3600)*pi*t/hours)'
  vars = 'kappa rho shortwave cp hours'
  vals = '1.0 1.0 1.0 1.0 1.0'
[]
[mms_exact]
  type = ParsedFunction
  value = 'x*y*exp(-3.08641975308642e-5*t)'
[]

!---

### Temporal Convergence: Input

> A third input is created. The content is follows the input to the spatial counter part, as such is

Reviewer: ```suggestion
A third input is created. The content follows the input to the spatial counter part and as such is
```

included completely below.

!listing tutorial03_verification/step04_mms/2d_mms_temporal.i link=false

!---

### Temporal Convergence: Run

> The `mms` package can be used to perform the convergence study. In this case, both a first and

Reviewer: ```suggestion
The `mms` package can be used to perform the convergence study. In this case, both first and
```
